pygmt/xarray_backend.py

```diff
@@ -0,0 +1,120 @@
+"""
+An xarray backend for reading raster grid/image files using the 'gmt' engine.
+"""
+
+import os
+from pathlib import Path
+from typing import Literal
+
+import xarray as xr
+from pygmt.clib import Session
+from pygmt.exceptions import GMTInvalidInput
+from pygmt.helpers import build_arg_list
+from pygmt.src.which import which
+from xarray.backends import BackendEntrypoint
+
+
+class GMTBackendEntrypoint(BackendEntrypoint):
+    """
+    Xarray backend to read raster grid/image files using 'gmt' engine.
+
+    Internally, GMT uses the NetCDF C library to read NetCDF files, and GDAL for GeoTIFF
+    files.
+
+    When using :py:func:`xarray.open_dataarray` or :py:func:`xarray.load_dataarray` with
+    ``engine='gmt'``, pass the ``decode_kind`` parameter that can be either:
+
+    - ``grid`` - for reading single-band raster grids
+    - ``image`` - for reading multi-band raster images
+
+    Examples
+    --------
+    Read a single-band NetCDF file using ``decode_kind="grid"``
+
+    >>> import pygmt
+    >>> import xarray as xr
+    >>>
+    >>> da_grid: xr.Dataset = xr.open_dataarray(
+    ...     "@static_earth_relief.nc", engine="gmt", decode_kind="grid"
+    ... )
+    >>> da_grid
+    <xarray.DataArray 'z' (lat: 14, lon: 8)> Size: 448B
+    [112 values with dtype=float32]
+    Coordinates:
+      * lat      (lat) float64 112B -23.5 -22.5 -21.5 -20.5 ... -12.5 -11.5 -10.5
+      * lon      (lon) float64 64B -54.5 -53.5 -52.5 -51.5 -50.5 -49.5 -48.5 -47.5
+    Attributes:
+        Conventions:   CF-1.7
+        title:         Produced by grdcut
+        history:       grdcut @earth_relief_01d_p -R-55/-47/-24/-10 -Gstatic_eart...
+        description:   Reduced by Gaussian Cartesian filtering (111.2 km fullwidt...
+        actual_range:  [190. 981.]
+        long_name:     elevation (m)
+
+    Read a multi-band GeoTIFF file using ``decode_kind="image"``
+
+    >>> da_image: xr.Dataset = xr.open_dataarray(
+    ...     "@earth_night_01d", engine="gmt", decode_kind="image"
+    ... )
+    >>> da_image
+    <xarray.DataArray 'z' (band: 3, y: 180, x: 360)> Size: 194kB
+    [194400 values with dtype=uint8]
+    Coordinates:
+      * y        (y) float64 1kB 89.5 88.5 87.5 86.5 ... -86.5 -87.5 -88.5 -89.5
+      * x        (x) float64 3kB -179.5 -178.5 -177.5 -176.5 ... 177.5 178.5 179.5
+      * band     (band) uint8 3B 1 2 3
+    Attributes:
+        long_name:  z
+    """
+
+    description = "Open raster (.grd, .nc or .tif) files in Xarray via GMT."
+    open_dataset_parameters = ("filename_or_obj", "decode_kind")
+    url = "https://github.com/GenericMappingTools/pygmt"
+
+    def open_dataset(  # type: ignore[override]
+        self,
+        filename_or_obj: str | os.PathLike,
+        *,
+        drop_variables=None,  # noqa: ARG002
+        decode_kind: Literal["grid", "image"],
```

Comment:
Why choose the name `decode_kind`? The original [`xr.open_dataarray`](https://docs.xarray.dev/en/stable/generated/xarray.open_dataarray.html#xarray-open-dataarray) method has parameters like `decode_cf`/`decode_times`/`decode_timedelta`, but here `decode` implies reading metadata from netCDF and converting to meaningful Python objects (e.g., datetime64). Maybe `raster_kind` is a better name?

Reply:
Yeah, it does sound like `decode_` is more decoding CF metadata variables, especially after reading https://github.com/pydata/xarray/issues/8548 where they're talking about `encode` and `decode` methods. Lemme rename `decode_kind` to `raster_kind` then.